File: psk/bpsk.m

Fs = 1e6;
bits = logical(randi([0, 1], 1, 10));

symbols = mapper(bits);
signal = pulse_shaping(symbols);

t = 0:length(bits)/(length(signal)-1):((length(signal)-1)/(length(signal)-1))*length(bits);

plot(t, repelem(bits, 32), "DisplayName", "Bits", "LineWidth", 1); hold on;
plot(t, signal,"DisplayName", "BPSK", "LineWidth", 1);
legend;
grid on;
ylim([-2 2])

function symbols = mapper(bits)
    I = bits * -2 + 1;
    symbols = complex(I, 0);
end

function signal = pulse_shaping(symbols)
    N = length(symbols);
    step = N*32;
    t = 0:N/step:N-N/step;
    phase = (repelem(symbols, 32) < 0) * pi + pi/2;
    signal = cos(2*pi*t + phase);
end

File: dpsk/bpsk.py
import numpy as np
import matplotlib.pyplot as plt

Fs = 1
bits = np.random.randint(0, 2, 10).astype(bool)

def mapper(bits):
    I = bits * -2 + 1
    return I.astype(complex)

def pulse_shaping(symbols):
    N = len(symbols)
    samples_per_symbol = 32
    step = N * samples_per_symbol
    t = np.linspace(0, N-N/step, step)
    phase = np.where(np.repeat(symbols, samples_per_symbol) < 0, np.pi, 0) + np.pi/2
    signal = np.cos(2 * np.pi * t + phase)
    return signal, t

symbols = mapper(bits)
signal, t = pulse_shaping(symbols)

t_bits = np.linspace(0, len(bits), len(signal), endpoint=False)
bits_rep = np.repeat(bits, 32).astype(float)

plt.figure(figsize=(10, 4))
plt.plot(t_bits, bits_rep, label="Bits", linewidth=1)
plt.plot(t, signal, label="BPSK", linewidth=1)
plt.legend()
plt.grid(True)
plt.ylim([-2, 2])
plt.title("BPSK Signal and Bits")
plt.xlabel("Time")
plt.ylabel("Amplitude")
plt.show()
